### PR TITLE
Merge program counter into stack and add back the End instruction.

### PR DESCRIPTION
**Pull request by q82419**
1. Merged `InstrProvider` into `StackManager`.
2. Added back the `End` instruction at the end of expressions.
3. Refine the validator workflow.

### CODE REVIEW

File: include/runtime/stackmgr.h [hunk 1]
```diff
@@ -98,20 +101,19 @@ class StackManager {
   }
 
   /// Unsafe pop top frame. Return number of popped label.
```
Comment (hydai) [inline include/runtime/stackmgr.h:103]: Your comment doesn't fit the code.

File: include/runtime/stackmgr.h [hunk 2]
```diff
@@ -127,6 +129,24 @@ class StackManager {
   /// Unsafe leave top label.
   void leaveLabel() { LabelStack.pop_back(); }
 
+  /// Unsafe getter of next instruction to execute.
+  const AST::Instruction *getNextInstr() {
+    if (LabelStack.size() == 0) {
+      return nullptr;
+    }
+    auto &L = LabelStack.back();
+    return (*(L.Curr++)).get();
```
Comment (hydai) [inline include/runtime/stackmgr.h:138]: You use `++` without checking the boundary.

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/runtime/stackmgr.h (hunk 2 above)
Comment (q82419) [reply include/runtime/stackmgr.h:138]: This is unsafe version, which bases on there's always an End instruction before the end of instruction vector.

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**